### PR TITLE
T16975

### PR DESCRIPTION
**Pull request by smspillaz**
Essentially this wraps a D-Bus method call by forwarding all of
its in-arguments and back-proxying its return value or error
to the caller, with the side effect of showing a splash screen
for the application if the application wasn't open yet. We'll
use this in DiscoveryFeed to ensure that we show a splash screen
for discovery-feed applications when they are opened to a given
page from the feed itself.

https://phabricator.endlessm.com/T16975

### CODE REVIEW

File: js/ui/shellDBus.js [hunk 1]
```diff
@@ -592,26 +612,50 @@ const AppLauncher = new Lang.Class({
...
+            invocation.return_error_literal(Gio.IOErrorEnum,
+                                            Gio.IOErrorEnum.NOT_FOUND,
+                                            'Unable to launch app ' + appName + ': Not installed');
+            return;
+        }
+
+        activationContext.activateViaDBusCall(busName, path, interfaceName, method, args, function(error, result) {
+            if (error) {
+                logError(error);
+                invocation.return_error_literal(Gio.IOErrorEnum,
```
Comment (cosimoc) [inline js/ui/shellDBus.js]: Can you simply return the error here instead of reconstructing it?

File: js/ui/shellDBus.js [hunk 2]
```diff
@@ -592,24 +604,23 @@ const AppLauncher = new Lang.Class({
     },
 
     LaunchAsync: function(params, invocation) {
-        if (name == 'eos-app-store') {
+        let [appName, timestamp] = params;
+
+        if (appName == 'eos-app-store') {
             Main.appStore.show(timestamp, true);
+            Main.appStore.appLaunched = true;
```
Comment (smspillaz) [inline js/ui/shellDBus.js]: My feeling here is that the old behaviour is wrong - there's no reason to touch Main.appStore.appLaunched if we're launching some other app. That said, I'll separate it out.

File: js/ui/appActivation.js [hunk 3]
```diff
@@ -90,6 +90,48 @@ const AppActivationContext = new Lang.Class({
...
+            try {
+                bus = Gio.bus_get_finish(result);
+            } catch (error) {
+                done(error, null);
+                return;
+            }
+
+            let proxy = new Gio.DBusProxy({
+                g_bus_type: Gio.BusType.SESSION,
+                g_connection: bus,
```
Comment (smspillaz) [inline js/ui/appActivation.js]: Ah okay, that's handy. I'll look into doing that.

Comment (smspillaz) [inline js/ui/appActivation.js]: Damn, I had a try of this and it doesn't seem to work :( I've left the old code in place for now.

File: js/ui/shellDBus.js (hunk 1 above)
Comment (smspillaz) [inline js/ui/shellDBus.js]: I don't think so in the sense that we need to convert the error into a GIOError (and should probably wrap it too.

**Comment by smspillaz** [comment]
Comments addressed - I also fixed a bug where we were incorrect wrapping return values where we didn't need to.

File: js/ui/shellDBus.js (hunk 2 above)
Comment (cosimoc) [inline js/ui/shellDBus.js]: It does not look like you separated it out?

File: js/ui/shellDBus.js [hunk 4]
```diff
@@ -607,7 +615,6 @@ const AppLauncher = new Lang.Class({
 
         if (appName == 'eos-app-store') {
             Main.appStore.show(timestamp, true);
-            Main.appStore.appLaunched = true;
```
Comment (cosimoc) [inline js/ui/shellDBus.js]: Oh, I see that you undo the change in this commit. Please squash it instead with the previous commit, so that none of the two changes this behavior.

Comment (smspillaz) [inline js/ui/shellDBus.js]: Oops, yeah, this revert needed to go into the other commit. Fixed it now.

**Comment by smspillaz** [comment]
I also added a third commit for what I understand to be the correct behaviour of appLaunched.
